### PR TITLE
Implementing the User Management API

### PR DESCRIPTION
**Pull request by hiranya911**
* Introduces a set of new methods to `FirebaseAuth` to create and manage user accounts.
* Introduces new public types `User`, `UserMetadata` and `ProviderUserInfo` for passing around user data.
* New unit and integration tests to cover the new user cases.

Related issue: #33

### CODE REVIEW

**Comment by hiranya911** [comment]
The API for this is still being reviewed, but I think it's stable enough to start a code review. I don't expect major changes except for perhaps a trivial renaming of couple of things.

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+   * @return a user ID string or null.
+   */
+  public String getUid() {
+    return uid;
+  }
+
+  /**
+   * Returns the user's display name at the IdP
+   *
+   * @return a display name string or null.
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Do we use @Nullable annotations in the admin SDK?

File: src/main/java/com/google/firebase/auth/User.java [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+  }
+
+  private static void checkEmail(String email) {
+    checkArgument(!Strings.isNullOrEmpty(email), "email cannot be null or empty");
+    checkArgument(email.matches("^[^@]+@[^@]+$"));
+  }
+
+  private static void checkPassword(String password) {
+    checkArgument(!Strings.isNullOrEmpty(password), "password cannot be null or empty");
+    checkArgument(password.length() >= 6, "password must be at least 6 characters long");
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/User.java]: FWIW, it seems kind of strange to have this in the SDK; I'd be inclined torely on the backend to enforce any applicable password restrictions...  and I worry that the backend might have other requirements which could manifest slightly differently which seems undesirable.

File: src/main/java/com/google/firebase/auth/User.java [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+    }
+
+    Map<String, Object> build() {
+      return ImmutableMap.copyOf(properties);
+    }
+  }
+
+  /**
+   * A class for updating the attributes of an existing user. An instance of this class can be
+   * obtained via a {@link User} object, or from an user ID string. Specify the changes to be
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/User.java]: "a user ID"

File: src/main/java/com/google/firebase/auth/User.java [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+     *
+     * @param displayName a display name string or null
+     */
+    public Updater setDisplayName(String displayName) {
+      properties.put("displayName", displayName);
+      return this;
+    }
+
+    /**
+     * Updates the Photo URL of this user. Calling this method with null a null argument removes
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/User.java]: null a null?

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 5]
```diff
@@ -62,10 +69,11 @@ private FirebaseAuth(FirebaseApp firebaseApp) {
    */
   @VisibleForTesting
   FirebaseAuth(
-      FirebaseApp firebaseApp, GooglePublicKeysManager googlePublicKeysManager, Clock clock) {
+      final FirebaseApp firebaseApp, GooglePublicKeysManager googlePublicKeysManager, Clock clock) {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseAuth.java]: I think it's some left over change from a prototyping work I was doing. Removed.

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 6]
```diff
@@ -195,6 +203,110 @@ public FirebaseToken then(@NonNull Task<String> task) throws Exception {
             });
   }
 
+  /**
+   * Gets the user data corresponding to the specified user ID.
+   *
+   * @param uid A user ID string.
+   * @return A {@link Task} which will complete successfully with a {@link User} instance.
+   *     If an error occurs while retrieving user data or if the specified user ID does not exist,
+   *     the task fails by throwing a FirebaseAuthException.
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseAuth.java]: Done

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.auth;
+
+import com.google.firebase.auth.internal.GetAccountInfoResponse;
+
+/**
+ * Contains metadata regarding how a user is known by a particular identity provider (IdP).
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Done

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  ProviderUserInfo(GetAccountInfoResponse.Provider response) {
+    this.uid = response.getUid();
+    this.displayName = response.getDisplayName();
+    this.email = response.getEmail();
+    this.photoUrl = response.getPhotoUrl();
+    this.providerId = response.getProviderId();
+  }
+
+  /**
+   * Returns the user's unique ID at the IdP.
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Done

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java (hunk 1 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Added both here and at `User`

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+   *
+   * @return a URL string or null.
+   */
+  public String getPhotoUrl() {
+    return photoUrl;
+  }
+
+  /**
+   * Returns the ID of the identity provider. This can be a short domain name (e.g. google.com) or
+   * the OP identifier of an OpenID IdP.
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Stands of "OpenID Provider". Changed the text to explain it.

File: src/main/java/com/google/firebase/auth/User.java (hunk 2 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/User.java]: Node Admin SDK does this. I'll try to figure out if it's needed at the SDK-end.

File: src/main/java/com/google/firebase/auth/User.java (hunk 3 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/User.java]: Fixed

File: src/main/java/com/google/firebase/auth/User.java (hunk 4 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/User.java]: Fixed

**Comment by hiranya911** [comment]
I'll merge this as it is. If any more changes come up after the final API review, I'll send a new PR.

**Pull request merged**